### PR TITLE
Dima

### PR DESCRIPTION
**Pull request by DzmitryI**

### CODE REVIEW

File: src/components/UI/input/input.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+
+const Input = (props) => {
+  const inputType = props.type || 'text';
+  const htmlFor = inputType + Math.random();
```
Comment (maximsan) [inline src/components/UI/input/input.js]: Why do you need Math.random() here ?

File: src/components/memberTasksGrid/memberTasksGrid.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import React, { Component } from 'react';
+
+export default class MemberTasksGrid extends Component {
```
Comment (maximsan) [inline src/components/memberTasksGrid/memberTasksGrid.js]: It's preferable to use stateless component

File: src/components/membersGrid/membersGrid.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    itemList: null,
+    members: [],
+    loading: true,
+  };
+
+  fetchService = new FetchService();
+
+  async componentDidMount() {
+    const getData = this.fetchService.getAllMember;
+    const members = await getData();
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: We can do it in one line

File: src/components/membersGrid/membersGrid.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      members,
+      loading: false,
+    });
+  }
+
+  countAge(value) {
+    const curDate = new Date();
+    const birthDate = new Date(value);
+    const age = curDate.getFullYear() - birthDate.getFullYear();
+    return curDate.setFullYear(1972) < birthDate.setFullYear(1972) ? age - 1 : age;
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: What is the magic number ?)

File: src/components/membersGrid/membersGrid.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  onChangeClick = ({ target }) => {
+    const editMemberId = target.closest('tr').id;
+    const curMembers = [...this.state.members].filter((el) => el.value[0] === editMemberId);
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: I believe that filer does  not mutate data

File: src/components/membersGrid/membersGrid.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      await getData(delMemberId);
+    } catch (err) {
+      alert(err); // TypeError: failed to fetch
+    }
+
+    // const members = await getData(delMember);
+    // this.setState({
+    //   members,
+    //   loading: false,
+    // });
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: remove commented code

File: src/components/membersGrid/membersGrid.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    const birthDate = new Date(value);
+    const age = curDate.getFullYear() - birthDate.getFullYear();
+    return curDate.setFullYear(1972) < birthDate.setFullYear(1972) ? age - 1 : age;
+  }
+
+  onChangeClick = ({ target }) => {
+    const editMemberId = target.closest('tr').id;
+    const curMembers = [...this.state.members].filter((el) => el.value[0] === editMemberId);
+    const [{ value }] = curMembers;
+    if (target.tagName === 'SPAN') {
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: I'm sure you can make it better, this comparison looks strange.

File: src/components/membersGrid/membersGrid.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    } else {
+      this.props.onRegisterClick(value, 'Detail Member page');
+    }
+  };
+
+  onDeleteClick = async ({ target }) => {
+    const delMemberId = target.closest('tr').id;
+    const curMembers = [...this.state.members];
+    this.setState({ members: curMembers.filter((el) => el.value[0] !== delMemberId) });
+    const getData = this.fetchService.delMember;
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: why do not do in in one line ?

File: src/components/membersGrid/membersGrid.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      this.props.onRegisterClick(value, 'Edit Member page');
+    } else {
+      this.props.onRegisterClick(value, 'Detail Member page');
+    }
+  };
+
+  onDeleteClick = async ({ target }) => {
+    const delMemberId = target.closest('tr').id;
+    const curMembers = [...this.state.members];
+    this.setState({ members: curMembers.filter((el) => el.value[0] !== delMemberId) });
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: Why value[0] is not just id ?

File: src/components/membersGrid/membersGrid.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    const getData = this.fetchService.delMember;
+    try {
+      await getData(delMemberId);
+    } catch (err) {
+      alert(err); // TypeError: failed to fetch
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: remove redundant comment & create a good error handler not just alert. for example set the error in state and show as a props of custom Error/Alert component

File: src/components/membersGrid/membersGrid.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  render() {
+    const { members, loading } = this.state;
+    const { onRegisterClick, onTaskClick, isOpen } = this.props;
+    if (loading) {
+      return <Spinner />;
+    }
+    const cls = ['members-wrap'];
+    if (isOpen) {
+      cls.push('close');
+    }
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: i use the same approach several times, I guess it's better to create smth common for it or review and simplify it

File: src/components/membersGrid/membersGrid.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    return (
+      <div className={cls.join(' ')}>
+        <h1>Members Manage Grid</h1>
+        <button type='button' className='btn btn-register' onClick={() => onRegisterClick([], 'Create Member page')}>
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: Avoid to use inline callbacks in code

Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: Do we really need to pass empty array as a first param to callback? it looks redundant.

File: src/components/membersGrid/membersGrid.scss [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  padding: 0;
+}
+
+.btn-register {
+  width: 120px;
+  height: 40px;
+}
+
+.btn-progress,
+.btn-tasks,
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.scss]: btn-tasks seems like an odd name. better to name it with using verb..

File: src/page/memberPage/memberPage.js [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+      return true;
+    }
+    let isValid = true;
+    if (validation.required) {
+      isValid = value.trim() !== '';
+    }
+    return isValid;
+  }
+
+  isInValid(valid, touched, shouldValidation) {
```
Comment (maximsan) [inline src/page/memberPage/memberPage.js]: It seems like unused method

File: src/page/memberPage/memberPage.js [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+    return !valid && touched && shouldValidation;
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.member !== prevProps.member && this.props.member.length !== 0) {
```
Comment (maximsan) [inline src/page/memberPage/memberPage.js]: this.props.member !== prevProps.member && this.props.member.length is the same

File: src/page/memberPage/memberPage.js [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+                  Male
+                </option>
+                <option value='sex2'>Female</option>
+              </select>
+            </div>
+            <div className='form-group row'>
+              <button
+                className='btn btn-add'
+                disabled={!this.state.isFormValid}
+                type='submin'
```
Comment (maximsan) [inline src/page/memberPage/memberPage.js]: typo

File: src/services/fetch-service.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    return await res.json();
+  };
+
+  getAllMember = async () => {
+    const res = await this.getResource(`/UserProfile.json`);
+    const members = [];
+    Object.entries(res).forEach((key) => {
+      members.push({
+        value: key,
```
Comment (maximsan) [inline src/services/fetch-service.js]: What is value exactly here? Maybe it better to use computer props ?

File: src/services/fetch-service.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  };
+
+  editMember = async (memberId, body) => {
+    return await this.editResource(`/UserProfile/${memberId}`, body);
+  };
+
+  delResource = async (url) => {
+    await fetch(`${this.API_BASE}${url}`, {
+      method: 'DELETE',
+    }).then((response) => response.json());
```
Comment (maximsan) [inline src/services/fetch-service.js]: From my point of view It's better to return boolean or string answer from here to be 100% sure it was deleted or text of  error

**Comment by maximsan** [comment]
I will approve it, but you obviously have to fix all notes. Thanks!

File: src/components/UI/input/input.js (hunk 1 above)
Comment (DzmitryI) [reply src/components/UI/input/input.js]: fixed

File: src/components/membersGrid/membersGrid.js (hunk 3 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: fixed

File: src/components/membersGrid/membersGrid.js (hunk 4 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: fixed

File: src/components/membersGrid/membersGrid.js (hunk 6 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: fixed

File: src/components/membersGrid/membersGrid.js (hunk 7 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: ~

File: src/components/membersGrid/membersGrid.js (hunk 9 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: fixed

**Pull request merged**